test/e2e/nodeconditions.go

```diff
@@ -0,0 +1,249 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	mb = 1024 * 1024
+	gb = 1024 * mb
+
+	// TODO(madhusudancs): find a way to query kubelet's disk space manager to obtain this value. 256MB
+	// is the default that is set today. This test might break if the default value changes. This value
+	// can be configured by setting the "low-diskspace-threshold-mb" flag while starting a kubelet.
+	// However, kubelets are started as part of the cluster start up, once, before any e2e test is run,
+	// and remain unchanged until all the tests are run and the cluster is brought down. Changing the
+	// flag value affects all the e2e tests. So we are hard-coding this value for now.
+	lowDiskSpaceThreshold uint64 = 256 * mb
+
+	nodeconditionsOODTimeOut = 1 * time.Minute
+
+	numNodeConditionsPods = 3
+
+	unfilledNodeIdx = 0
+)
+
+type podId struct {
+	name  string
+	value string
+}
+
+// Plan:
+// 1. Fill disk space on all nodes except one. One node is left out so that we can schedule pods
+//    on that node. Arbitrarily choose that node to be node with index 0.
+// 2. Get the CPU capacity on unfilled node.
+// 3. Divide it into number of pods we want to schedule - 1. We want to schedule 3 pods, so
+//    divide CPU capacity by 2.
+// 4. Request the divided capacity for each pod.
+// 5. Attempt to schedule the final pod and observe that it fails.
+//
+// Reason: although there is CPU available on other nodes, pods are not schedulable on those nodes
+//         because the other nodes are out of disk.
+//
+// TODO(madhusudancs): Today, we use CPU instead of directly using storage as resource constraint
```

Comment:
Once we have local disk storage as a schedulable resource, we actually won't need this feature at all, because the scheduler will prevent scheduling pods onto nodes where free disk space < requested disk space. (We'll still want the OOD nodeStatus, for debugging purposes, but there won't be any need for it to feed into the scheduler anymore.) So this TODO is not something we would do.